protocols/gossipsub/src/metrics.rs

```diff
@@ -40,6 +40,7 @@ const DEFAULT_MAX_TOPICS: usize = 300;
 // store metrics.
 const DEFAULT_MAX_NEVER_SUBSCRIBED_TOPICS: usize = 50;
 
+#[derive(Debug, Clone, Copy)]
```

Comment:
Copy is a rather strict promise, I’d only add it where there is a clear benefit that would justify some pain to keep it (like for PeerId). For `*Config` structures it can easily happen that some String or other heap allocated structure is added in the future, which would then be a larger breaking change (field addition can for users be done in a mostly compatible fashion using `default` and `with_x()` functions).

Reply:
Indeed. On a similar note I saw that some configs have `#[non_exhaustive]` on them and others do not. Maybe it is related to the fact that some are considered to be "done" and some are still evolving, but I found it interesting as well.

Reply:
I think it is okay to make that promise now and perhaps remove it again later. Yes it is a breaking change but it is

a) easily fixed
b) may not break users in all circumstances (only if they use `Copy`)
c) we make minor releases (which are breaking < `0.x`) all the time

Hence I think it is okay to give users you may benefit from this the benefits now and accept it if we need to break it again later.

@mxinden Thoughts?

Reply:
Is there much value in offering `Copy` for this particular config? In other words, I would suspect downstream code to only need to clone it once or twice, thus I don't see the benefit of offering `Copy` here.

> Indeed. On a similar note I saw that some configs have #[non_exhaustive] on them and others do not. Maybe it is related to the fact that some are considered to be "done" and some are still evolving, but I found it interesting as well.

Correct. This is a pattern used in the past. I personally don't see much value in it. As a downstream user I would prefer a breaking change over accidentally missing some new config option or event variant.